enos/modules/vault_agent/main.tf

```diff
@@ -0,0 +1,69 @@
+terraform {
+  required_providers {
+    aws = {
+      source = "hashicorp/aws"
+    }
+    enos = {
+      source = "app.terraform.io/hashicorp-qti/enos"
+    }
+  }
+}
+
+variable "vault_agent_template_destination" {
+  type        = string
+  description = "The destination of the template rendered by Agent"
+}
+
+variable "vault_agent_template_contents" {
+  type        = string
+  description = "The template contents to be rendered by Agent"
+}
+
+variable "vault_root_token" {
+  type        = string
+  description = "The Vault root token"
+}
+
+variable "vault_instances" {
+  type = map(object({
+    private_ip = string
+    public_ip  = string
+  }))
+  description = "The Vault cluster instances that were created"
+}
+
+variable "vault_instance_count" {
+  type        = number
+  description = "How many vault instances are in the cluster"
+}
+
+variable "vault_install_dir" {
+  type        = string
+  description = "The directory where the Vault binary will be installed"
+}
+
+locals {
+  vault_instances = {
+    for idx in range(var.vault_instance_count) : idx => {
+      public_ip  = values(var.vault_instances)[idx].public_ip
+      private_ip = values(var.vault_instances)[idx].private_ip
+    }
+  }
+}
+
+resource "enos_remote_exec" "set_up_approle_auth_and_agent" {
+  content = templatefile("${path.module}/templates/set-up-approle-and-agent.sh", {
+    vault_install_dir       = var.vault_install_dir
+    vault_local_binary_path = "${var.vault_install_dir}/vault"
+    vault_token             = var.vault_root_token
+    vault_agent_template_destination = var.vault_agent_template_destination
+    vault_agent_template_contents = var.vault_agent_template_contents
+    vault_instances   = jsonencode(local.vault_instances)
+  })
+
+  transport = {
+    ssh = {
+      host = local.vault_instances[0].public_ip
```

Comment:
Dumb question: how do I actually do this?

I've tried passing the following in from the scenario into the main.tf, and both seem to fail for different reasons.

```
vault_instance = step.create_vault_cluster.vault_instances[0].public_ip
// fails with:
// module.create_vault_cluster.vault_instances is object with 3 attributes
//│
//│ The given key does not identify an element in this collection value. An
//│ object only supports looking up attributes by name, not by numeric index.
vault_instance = values(step.create_vault_cluster.vault_instances)[0].public_ip
// fails with │ This object does not have an attribute named "vault_instances"
```

I'm admittedly a little confused as to whether any black magic is occurring in the examples where it takes the variable and turns it into a local. I assume there's something I'm missing, but most examples send the full instance list to the main.tf.

Reply:
if you move that code into a local it might work..., i.e.:
```
locals {
...
    agent_instance = step.create_vault_cluster.vault_instances[0].public_ip
...
}
```

Actually, I don't think you need to wrap the call with the `values()` function. Might want to try that first.

Reply:
Hmm, I apparently can't use locals in this file:
```
│ Error: Unsupported block type
│
│   on /Users/violet/Repositories/vault/enos/enos-scenario-agent.hcl line 154, in scenario "agent":
│  154:     locals {
│
│ Blocks of type "locals" are not expected here.
```

Reply:
You already have a `locals` declared on line 17 of the scenario file, you should be able to put it there. If it doesn't work, it would be because it references a step and if that's the case then I guess sticking with what you had that works would be your only option.

Reply:
Hmm, yeah it can't reference a step, so I guess I need to keep it as-is. I guess this approach keeps it inline with the other modules, at least. Thanks for helping me work through this regardless!